### ISSUE TITLE
DrudeForce: Particle index is used by two different Drude particles

### ISSUE THREAD

**Issue by dspoel** [openmm/openmm#3977]
Sorry if this is a FAQ. We are trying to forge some OpenMM input using Drude particles but without repeating the same information more than necessary. For instance, we noted that in the Charmm polarizable force field identical particles such as the drudes on an ASP side chain oygens have different types rude-ASP-OD1/2 but their parameters are identical. Cannot this be avoided using classes? 

In the attached example we tried to model a single water molecule with two classes for the drudes rather than three types.
```
  <DrudeForce>
    <Particle class1="o3_b" class2="o3_s_b" polarizability="0.0004" charge="-1.52796" thole="0"/>
    <Particle class1="h_b" class2="h_s_b" polarizability="0.0001" charge="-0.928007" thole="0"/>
  </DrudeForce>
```

However, when running this example using
```
./run_openmm_simulation.py -pdb water.pdb -xml WAT.xml -dat datel.dat  -bonds cubic
```
we end up with this error, suggesting assignment of drudes to atoms is going wrong:
```
openmm.OpenMMException: DrudeForce: Particle index is used by two different Drude particles: 5
```

Is this a bug that can be fixed without duplicating tons of data?

[openmm_issue.zip](https://github.com/openmm/openmm/files/10817094/openmm_issue.zip)

**Comment by peastman** [comment]
> Cannot this be avoided using classes?

In principle yes.  The challenge is deciding what should be merged.  The XML files in OpenMM were automatically generated from the corresponding CHARMM files.  For example, you mentioned the oxygens in ASP:

```xml
    <Type name="ASP-OD1" class="OD2C2A" mass="15.9994" element="O"/>
    <Type name="Drude-ASP-OD1" class="DRUD" mass="0.0"/>
    <Type name="ASP-OD2" class="OD2C2A" mass="15.9994" element="O"/>
    <Type name="Drude-ASP-OD2" class="DRUD" mass="0.0"/>
```

They were generated based on these lines in toppar_drude_master_protein_2019h.str:

```
ATOM OD1  OD2C2A   0.003  ALPHA -0.699  THOLE 2.399 
ATOM OD2  OD2C2A   0.003  ALPHA -0.699  THOLE 2.399 
```

Since the CHARMM file defines the Drude particles separately, they get translated as separate definitions.  Of course, they have identical parameters and so in principle they could be merged together.  But there are also other Drude particles in other residues that also have those same parameters.  Should we merge all of them, or is that going too far?  How should we decide which to merge?

> Is this a bug that can be fixed without duplicating tons of data?

Something seems to be messed up in how the system is getting built.  Let me investigate.

**Comment by dspoel** [comment]
If the Charmm force field works as it should there is no reason to change those. But my group would like to define models without duplicating information, so it would be appreciated if you could have a look!

**Comment by peastman** [comment]
There are a couple of problems in how you define the DrudeForce:

```xml
<DrudeForce>
    <Particle class1="o3_b" class2="o3_s_b" polarizability="0.0004" charge="-1.52796" thole="0"/>
    <Particle class1="h_b" class2="h_s_b" polarizability="0.0001" charge="-0.928007" thole="0"/>
</DrudeForce>
```

class1/type1 refers to the Drude particle, and class2/type2 refers to its parent atom.  Those lines tell it to treat the oxygen (with class "o3_b") as the Drude particle, and the particle with type "o3_s_b" as its parent atom.  They need to be reversed.

The other problem relates to why you really do need to use types instead of classes, at least for the `<DrudeForce>` tag.  There are two particles of class "h_s_b".  Those are going to be treated as Drude particles.  But which atom should each be attached to?  There's no way to know.  In this case, it actually ends up attaching both of them to the same hydrogen, leading to an error.

Here's a fixed version that works correctly.

```xml
  <DrudeForce>
    <Particle type1="o3_s_1" type2="o3_1" polarizability="0.0004" charge="-1.52796" thole="0"/>
    <Particle type1="h_s_1" type2="h_1" polarizability="0.0001" charge="-0.928007" thole="0"/>
    <Particle type1="h_s_2" type2="h_2" polarizability="0.0001" charge="-0.928007" thole="0"/>
  </DrudeForce>
```

That restriction only applies to the DrudeForce.  In your NonbondedForce and CustomNonbondedForce, it's fine to define the parameters by class instead of type.

**Comment by dspoel** [comment]
Thanks, but this is what I would like to prevent since we can end up with many identical types. A possibility would be to manually list the connections in the residue section, like bonds. Alternatively, another solution would be to turn off automatic drude generation and instead generate python code that calls DrudeForce.addParticle(). This would work as well, though somewhat more cumbersome at the user end.

**Comment by peastman** [comment]
Can you describe your project, and how the identical types cause problems?  What is the workflow for generating the XML file?

**Comment by dspoel** [comment]
We have our own force field and code that generates OpenMM input files in order to do simulations. The codes have not yet been released but will be in a short while. In the force field there is only one hydrogen atom type, for instance. We are planning to use OpenMM to modify force field parameters on the fly and it would  be cumbersome to first make many copies of the parameters, and then modify many parameters during a simulation. The plan is to implement weak coupling of force field parameters to observables along the lines of some older papers like this one. https://aip.scitation.org/doi/10.1063/1.476482

**Comment by peastman** [comment]
Can your code generate the XML files based on an internal description that has fewer parameters?  That could be done with a templating engine, or even just find and replace.  Your template might look like

```xml
  <DrudeForce>
    <Particle type1="o3_s_1" type2="o3_1" polarizability="O3_B_POLARIZABILITY" charge="O3_B_CHARGE" thole="O3_B_THOLE"/>
    <Particle type1="h_s_1" type2="h_1" polarizability="H_B_POLARIZABILITY" charge="H_B_CHARGE" thole="H_B_THOLE"/>
    <Particle type1="h_s_2" type2="h_2" polarizability="H_B_POLARIZABILITY" charge="H_B_CHARGE" thole="H_B_THOLE"/>
  </DrudeForce>
```

You could use `str.replace()` to fill in the correct parameter values.

**Comment by dspoel** [comment]
That is an interesting idea, however the problem is not generating the file, but rather manipulating the force field during the OpenMM simulation. In fact we have successfully run simulations using your earlier suggestion. How about my questions earlier, maybe you can just say whether or not those would be possible?

1.  to manually list the atom-drude connections in the residue section, like bonds, or
2. to turn off automatic drude generation and instead use python code that calls DrudeForce.addParticle()

**Comment by peastman** [comment]
> manipulating the force field during the OpenMM simulation.

I'm not sure that's a good idea.  The internal generator classes aren't part of the public API, so they often change between releases.  Editing them in memory does save the time of generating and parsing a new XML file, but you still have to create a new System either way, which will often take longer.

> to manually list the atom-drude connections in the residue section, like bonds, or

That would be a pretty big change to how it defines Drude particles.  I'd have to think about it, but I'm not sure it's a good idea.

> to turn off automatic drude generation and instead use Python code that calls DrudeForce.addParticle()

That is of course a possibility.  The ForceField doesn't do anything you can't do directly.  If you do that, be sure you set up exceptions and masses correctly.  Take a look at `DrudeGenerator.postprocessSystem()`.

**Comment by peastman** [comment]
> That means all cores will get at most one shell.

You might not get the right one, though.  You could end up with h_s_2 attached to H_3 and h_s_3 attached to H_2.  That wouldn't affect the simulation, but it might confuse your analysis.

**Comment by dspoel** [comment]
Indeed, the best solution would be to explicitly specify the core-drude pairs in the residue definition...
